### DISCUSSION TITLE
How am I supposed to store images in state for later use?

### DISCUSSION THREAD

**Discussion by Pugsworth** [emilk/egui#3510] [Q&A] [answered]
Let me preface this by saying I'm still very new to Rust. I understand the basics, but I definitely get lost with lifetimes.

I'm making a little GUI app to drag and drop images for some kind of processing. This requires keeping a reference to the image data via image::ImageBuffer. I then want to be able to display this image in the GUI at any time. Obviously, I cannot just use the Imagebuffer, so it needs to be converted to something egui understands.

Where I am getting stuck is this:
What struct am I supposed to keep a reference to? There's egui::Image, egui::Texture, egui::TextureId, egui::SizedTexture, etc...
I don't want to be building new instances of whatever I need to be able to draw it every frame.

I've been at this all day trying to figure it out so I wouldn't have to ask a potentially dumb question like this, but I'm at a loss now.

**Comment by JumpyLionnn** [comment, marked as answer]
Its pretty much a duplicate of this one https://github.com/emilk/egui/discussions/3431 .
If you want to load an image using the image crate this is the way ^
But just to let you know, using the image api in egui doesnt really load the image every frame, it caches it in the background

**Reply by Pugsworth** [reply]
Oh, thank you! I didn't find any other discussion posts that were exactly related to my question.

So if I'm getting this right, then actual loading of the texture would be deferred until the next "render" to utilize the ctx.load_texture?

This is what I came up with after looking through that discussion post.
```rust
// Load the image if output doesn't exist.
if let None = state.test.output {
    if let Some(img) = &state.test.input {
        let (w, h) = img.dimensions();
        let col_img = egui::ColorImage::from_rgb([w as usize, h as usize], img.to_rgb8().as_bytes());
        let handle = ctx.load_texture("test", col_img, TextureOptions::LINEAR);

        state.test.output = Some(handle);
    }
}

// Show the output image if it exists.
egui::CentralPanel::default()
.show(ctx, |ui| {
    if let Some(handle) = &state.test.output {
        let sized_tex = egui::load::SizedTexture::new(handle.id(), handle.size_vec2());
        let img = egui::Image::from_texture(sized_tex);
        ui.add(img);
    }
});
```

**Reply by JumpyLionnn** [reply]
It looks ok to me. does it work?
After creating the image widget you can use the normal egui image api: https://docs.rs/egui/latest/egui/widgets/struct.Image.html
This just a way to load it with the image crate if you need something from it.

**Reply by Pugsworth** [reply]
Yeah, doing this works just fine as far as I can tell.

The way I think tends to come from languages like C/C++ in which managing memory is paramount as well as pointer juggling. Needless to say, this is very weird to me. But, I'm getting there.

Thank you!
